packages/flame/lib/src/effects/controllers/effect_controller.dart

```diff
@@ -78,35 +86,115 @@ abstract class EffectController {
     double atMaxDuration = 0.0,
     double atMinDuration = 0.0,
   }) {
+    assert(
+      (duration ?? 1) >= 0,
+      'Duration cannot be negative: $duration',
+    );
+    assert(
+      (reverseDuration ?? 1) >= 0,
+      'Reverse duration cannot be negative: $reverseDuration',
+    );
+    assert(
+      (duration != null) || (speed != null),
+      'Either duration or speed must be specified',
+    );
+    assert(
+      !(duration != null && speed != null),
+      'Both duration and speed cannot be specified at the same time',
+    );
+    assert(
+      !(reverseDuration != null && reverseSpeed != null),
+      'Both reverseDuration and reverseSpeed cannot be specified at the '
+      'same time',
+    );
+    assert(
+      (speed ?? 1) > 0,
+      'Speed must be positive: $speed',
+    );
+    assert(
+      (reverseSpeed ?? 1) > 0,
+      'Reverse speed must be positive: $reverseSpeed',
+    );
+    assert(
+      !(infinite && repeatCount != null),
+      'An infinite effect cannot have a repeat count',
+    );
+    assert(
+      (repeatCount ?? 1) > 0,
+      'Repeat count must be positive: $repeatCount',
+    );
+    assert(
+      startDelay >= 0,
+      'Start delay cannot be negative: $startDelay',
+    );
+    assert(
+      atMaxDuration >= 0,
+      'At-max duration cannot be negative: $atMaxDuration',
+    );
+    assert(
+      atMinDuration >= 0,
+      'At-min duration cannot be negative: $atMinDuration',
+    );
     final isLinear = curve == Curves.linear;
-    final hasReverse = alternate || (reverseDuration != null);
+    final hasReverse =
+        alternate || (reverseDuration != null) || (reverseSpeed != null);
     final reverseIsLinear =
         reverseCurve == Curves.linear || ((reverseCurve == null) && isLinear);
+    final reverseHasDuration =
+        (reverseDuration != null) || (reverseSpeed == null && duration != null);
+
     final items = [
-      if (isLinear) LinearEffectController(duration),
-      if (!isLinear) CurvedEffectController(duration, curve),
-      if (atMaxDuration != 0) PauseEffectController(atMaxDuration, progress: 1),
+      // FORWARD
+      if (isLinear)
```

Comment:
All if-else statements should have an explicit body `{}`, I thought we had this in the analyzer already...

Reply:
This is inside the list constructor `[ ... ]`, so curly braces aren't even supported there...

Reply:
Ah right, it should be broken up into smaller parts then, so that this much logic isn't inside the list construction.

Reply:
Ok then